src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Completion/RemoteCompletionService.cs

```diff
@@ -0,0 +1,229 @@
+            return Response.CallHtml;
+        }
+
+        return Response.Results(mergedCompletionList);
+    }
+
+    private async ValueTask<VSInternalCompletionList?> GetCSharpCompletionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        int documentIndex,
+        Position mappedPosition,
+        TextEdit? provisionalTextEdit,
+        CompletionContext completionContext,
+        VSInternalClientCapabilities clientCapabilities,
+        RazorCompletionOptions razorCompletionOptions,
+        CancellationToken cancellationToken)
+    {
+        var generatedDocument = await remoteDocumentContext.Snapshot
+            .GetGeneratedDocumentAsync(cancellationToken).ConfigureAwait(false);
+        if (provisionalTextEdit is not null)
+        {
+            var generatedText = await generatedDocument.GetTextAsync(cancellationToken);
+            var startIndex = generatedText.GetPosition(provisionalTextEdit.Range.Start);
+            var endIndex = generatedText.GetPosition(provisionalTextEdit.Range.End);
+            var generatedTextWithEdit = generatedText.Replace(startIndex, length: endIndex - startIndex, newText: provisionalTextEdit.NewText);
+            generatedDocument = generatedDocument.WithText(generatedTextWithEdit);
+        }
+
+        // This is, to say the least, not ideal. In future we're going to normalize on to Roslyn LSP types, and this can go.
+        var options = new JsonSerializerOptions();
+        foreach (var converter in RazorServiceDescriptorsWrapper.GetLspConverters())
+        {
+            options.Converters.Add(converter);
+        }
+
+        if (JsonSerializer.Deserialize<RoslynCompletionContext>(JsonSerializer.SerializeToDocument(completionContext), options) is not { } roslynCompletionContext)
+        {
+            return null;
+        }
+
+        if (JsonSerializer.Deserialize<RoslynCompletionSetting>(JsonSerializer.SerializeToDocument(clientCapabilities.TextDocument?.Completion), options) is not { } roslynCompletionSetting)
+        {
+            return null;
+        }
+
+        var mappedLinePosition = mappedPosition.ToLinePosition();
+        var roslynCompletionList = await ExternalAccess.Razor.Cohost.Handlers.Completion.GetCompletionListAsync(
+            generatedDocument,
+            mappedLinePosition,
+            roslynCompletionContext,
+            clientCapabilities.SupportsVisualStudioExtensions,
+            roslynCompletionSetting,
+            cancellationToken);
+
+        if (roslynCompletionList is null)
+        {
+            // If we don't get a response from the delegated server, we have to make sure to return an incomplete completion
+            // list. When a user is typing quickly, the delegated request from the first keystroke will fail to synchronize,
+            // so if we return a "complete" list then the query won't re-query us for completion once the typing stops/slows
+            // so we'd only ever return Razor completion items.
```

Comment:
I feel that the above two spots should not actually happen and returning null there indicates unexpected state rather than expected "completion not ready". I'm adding assert for both of those cases to indicate that expectation more clearly.